<a href="https://colab.research.google.com/github/jaydenhurt1215-dotcom/CS4410-HW/blob/main/Jhurt_Hw_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Exercise 16.1 - Fashion MNIST

In [18]:
from tensorflow.keras.datasets import fashion_mnist
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

In [19]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


In [20]:
x_train = x_train.reshape((60000, 28, 28, 1))
x_test = x_test.reshape((10000, 28, 28, 1))

print(x_train.shape)
print(x_test.shape)

(60000, 28, 28, 1)
(10000, 28, 28, 1)


In [21]:
x_train=x_train.astype('float32')/255
x_test=x_test.astype('float32')/255

In [22]:
from tensorflow.keras.utils import to_categorical
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

print(y_train.shape)
print(y_test.shape)

(60000, 10)
(10000, 10)


In [23]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

cnn=Sequential()

cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(28,28,1)))
cnn.add(MaxPooling2D(pool_size=(2,2)))
cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2,2)))
cnn.add(Flatten())
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))



/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [24]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [25]:
cnn.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_4 (Conv2D)               │ (None, 26, 26, 64)     │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 13, 13, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 11, 11, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 5, 5, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 3200)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 128)            │       409,728 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 485,514 (1.85 MB)

 Trainable params: 485,514 (1.85 MB)

 Non-trainable params: 0 (0.00 B)

In [26]:
import time
start_time = time.perf_counter()
cnn.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

fashion_time=time.perf_counter()-start_time
print('Training Time:', fashion_time, 'seconds')

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 8s 7ms/step - accuracy: 0.8308 - loss: 0.4616 - val_accuracy: 0.8833 - val_loss: 0.3210
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8907 - loss: 0.3017 - val_accuracy: 0.8960 - val_loss: 0.2807
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9069 - loss: 0.2533 - val_accuracy: 0.9035 - val_loss: 0.2596
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9181 - loss: 0.2215 - val_accuracy: 0.8915 - val_loss: 0.2870
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9285 - loss: 0.1946 - val_accuracy: 0.9063 - val_loss: 0.2621
Training Time: 26.512334008999915 seconds


In [27]:
loss, fashion_accuracy = cnn.evaluate(x_test, y_test)
print('Test Accuracy:', fashion_accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9004 - loss: 0.2856
Test Accuracy: 0.9003999829292297


MNIST Comparison

In [28]:
from tensorflow.keras.datasets import mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

print (x_train.shape)
print (y_train.shape)
print (x_test.shape)
print (y_test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


In [29]:
x_train = x_train.reshape((60000, 28, 28, 1))
x_test = x_test.reshape((10000, 28, 28, 1))

x_train=x_train.astype('float32')/255
x_test=x_test.astype('float32')/255

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)


In [30]:
mnist_cnn=Sequential()
mnist_cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(28,28,1)))
mnist_cnn.add(MaxPooling2D(pool_size=(2,2)))

mnist_cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu'))
mnist_cnn.add(MaxPooling2D(pool_size=(2,2)))

mnist_cnn.add(Flatten())
mnist_cnn.add(Dense(units=128, activation='relu'))
mnist_cnn.add(Dense(units=10, activation='softmax'))

mnist_cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [31]:
start_time = time.perf_counter()
mnist_cnn.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

mnist_time=time.perf_counter()-start_time

print('Training Time:', mnist_time, 'seconds')

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 9s 8ms/step - accuracy: 0.9575 - loss: 0.1397 - val_accuracy: 0.9870 - val_loss: 0.0464
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9860 - loss: 0.0440 - val_accuracy: 0.9908 - val_loss: 0.0338
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9910 - loss: 0.0282 - val_accuracy: 0.9907 - val_loss: 0.0325
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9938 - loss: 0.0201 - val_accuracy: 0.9890 - val_loss: 0.0398
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9953 - loss: 0.0154 - val_accuracy: 0.9885 - val_loss: 0.0409
Training Time: 26.825105213999905 seconds


In [32]:
loss, mnist_accuracy = mnist_cnn.evaluate(x_test, y_test)
print('Test Accuracy:', mnist_accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9869 - loss: 0.0387
Test Accuracy: 0.9868999719619751


**Exercise 16.1 Results:** The model performed better on MNIST than Fashion-MNIST. The test accuracy for MNIST was 98.69%, while the test accuracy for Fashion-MNIST was 90.04%. The training times were very close. Fashion-MNIST took 26.51 seconds to train, while MNIST took 26.83 seconds.

Exercise 16.4 - Convnet Layers

In [33]:
no_dense_cnn=Sequential()
no_dense_cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(28,28,1)))
no_dense_cnn.add(MaxPooling2D(pool_size=(2,2)))
no_dense_cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu'))
no_dense_cnn.add(MaxPooling2D(pool_size=(2,2)))
no_dense_cnn.add(Flatten())
no_dense_cnn.add(Dense(units=10, activation='softmax'))

no_dense_cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [34]:
no_dense_cnn.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_8 (Conv2D)               │ (None, 26, 26, 64)     │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_8 (MaxPooling2D)  │ (None, 13, 13, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_9 (Conv2D)               │ (None, 11, 11, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_9 (MaxPooling2D)  │ (None, 5, 5, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_4 (Flatten)             │ (None, 3200)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 10)             │        32,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 106,506 (416.04 KB)

 Trainable params: 106,506 (416.04 KB)

 Non-trainable params: 0 (0.00 B)

In [35]:
start_time=time.perf_counter()
no_dense_cnn.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
no_dense_time=time.perf_counter()-start_time
print('Training Time:', no_dense_time, 'seconds')

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 9s 7ms/step - accuracy: 0.9478 - loss: 0.1710 - val_accuracy: 0.9848 - val_loss: 0.0538
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9846 - loss: 0.0493 - val_accuracy: 0.9880 - val_loss: 0.0409
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9888 - loss: 0.0362 - val_accuracy: 0.9897 - val_loss: 0.0363
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9914 - loss: 0.0273 - val_accuracy: 0.9898 - val_loss: 0.0365
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.9936 - loss: 0.0200 - val_accuracy: 0.9862 - val_loss: 0.0512
Training Time: 25.41313286700006 seconds


In [36]:
loss, no_dense_accuracy = no_dense_cnn.evaluate(x_test, y_test)
print('Test Accuracy:', no_dense_accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9858 - loss: 0.0452
Test Accuracy: 0.98580002784729


4096-Neuron Dense Layer

In [37]:
large_dense_cnn=Sequential()
large_dense_cnn.add(Conv2D(filters=64, kernel_size=(3,3), activation='relu', input_shape=(28,28,1)))
large_dense_cnn.add(MaxPooling2D(pool_size=(2,2)))
large_dense_cnn.add(Conv2D(filters=128, kernel_size=(3,3), activation='relu'))
large_dense_cnn.add(MaxPooling2D(pool_size=(2,2)))
large_dense_cnn.add(Flatten())
large_dense_cnn.add(Dense(units=4096, activation='relu'))
large_dense_cnn.add(Dense(units=128, activation='relu'))
large_dense_cnn.add(Dense(units=10, activation='softmax'))

large_dense_cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

/usr/local/lib/python3.13/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [38]:
large_dense_cnn.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_10 (Conv2D)              │ (None, 26, 26, 64)     │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_10 (MaxPooling2D) │ (None, 13, 13, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_11 (Conv2D)              │ (None, 11, 11, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_11 (MaxPooling2D) │ (None, 5, 5, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_5 (Flatten)             │ (None, 3200)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 4096)           │    13,111,296 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 128)            │       524,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 13,711,498 (52.31 MB)

 Trainable params: 13,711,498 (52.31 MB)

 Non-trainable params: 0 (0.00 B)

In [39]:
start_time=time.perf_counter()
large_dense_cnn.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.1)
large_dense_time=time.perf_counter()-start_time

print('Training Time:', large_dense_time, 'seconds')

Epoch 1/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 12s 11ms/step - accuracy: 0.9642 - loss: 0.1140 - val_accuracy: 0.9867 - val_loss: 0.0443
Epoch 2/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9874 - loss: 0.0398 - val_accuracy: 0.9868 - val_loss: 0.0423
Epoch 3/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9914 - loss: 0.0271 - val_accuracy: 0.9913 - val_loss: 0.0321
Epoch 4/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9946 - loss: 0.0175 - val_accuracy: 0.9917 - val_loss: 0.0366
Epoch 5/5
844/844 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.9948 - loss: 0.0171 - val_accuracy: 0.9915 - val_loss: 0.0323
Training Time: 42.93177699399985 seconds


In [40]:
loss, large_dense_accuracy = large_dense_cnn.evaluate(x_test, y_test)
print('Test Accuracy:', large_dense_accuracy)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9910 - loss: 0.0329
Test Accuracy: 0.9909999966621399


**Exercise 16.4 Results:** Removing the 1st Dense layer slightly decreased the test accuracy from 98.69% to 98.58%. It also decreased the training time from 26.83 seconds to 25.41 seconds. Adding a Dense layer with 4096-neurons increased the test accuracy to 99.10%, but it also increased the training time to 42.93 seconds. The 4096-neuron model had the highest accuracy, but it was also the slowest model to train.